File: tesserack-local-ai/app/src/lib/core/action-parser.js
// action-parser.js - Parse LLM responses
export const VALID_BUTTONS = new Set(['a', 'b', 'start', 'select', 'up', 'down', 'left', 'right']);
export const MAX_PLAN_ACTIONS = 12;

export function isValidButton(button) {
    return VALID_BUTTONS.has(String(button || '').toLowerCase());
}

/** Parse numbered candidate plans while pairing plans/actions by their number. */
export function parseMultiplePlans(response) {
    const plansByNumber = new Map();
    const actionsByNumber = new Map();

    for (const match of response.matchAll(/PLAN(\d+):\s*(.+?)(?=\n|$)/gi)) {
        plansByNumber.set(match[1], match[2].trim());
    }
    for (const match of response.matchAll(/ACTIONS(\d+):\s*(.+?)(?=\n|PLAN|$)/gi)) {
        actionsByNumber.set(match[1], match[2]
            .split(/[,\s]+/)
            .map(action => action.toLowerCase().trim())
            .filter(isValidButton)
            .slice(0, MAX_PLAN_ACTIONS));
    }

    const plans = [];
    for (const [number, plan] of plansByNumber) {
        const actions = actionsByNumber.get(number) || [];
        if (actions.length > 0) plans.push({ plan, actions });
    }

    if (plans.length === 0) {
        const fallback = parseResponse(response);
        if (fallback.actions.length > 0) plans.push(fallback);
    }
    return plans;
}

export function parseResponse(response) {
    console.log('[LLM RAW]', response.substring(0, 300));

    let plan = '';
    let actions = [];

    // Look for PLAN: line
    const planMatch = response.match(/PLAN:\s*(.+?)(?:\n|$)/i);
    if (planMatch) {
        plan = planMatch[1].trim();
    }

    // Look for ACTIONS: line
    const actionsMatch = response.match(/ACTIONS?:\s*(.+?)(?:\n|$)/i);
    if (actionsMatch) {
        const actionStr = actionsMatch[1].toLowerCase();
        const buttons = actionStr.split(/[,\s]+/);
        actions = buttons
            .map(b => b.trim())
            .filter(isValidButton)
            .slice(0, MAX_PLAN_ACTIONS);
    } else {
        // Tiny local models sometimes return only a raw button sequence even
        // when a labelled format was requested. Prefer using it over a blind fallback.
        actions = (response.toLowerCase().match(/\b(?:up|down|left|right|start|select|a|b)\b/g) || [])
            .filter(isValidButton)
            .slice(0, MAX_PLAN_ACTIONS);
    }

    // Fallback if no plan found
    if (!plan && actionsMatch) {
        const actionStart = response.search(/actions?:/i);
        if (actionStart > 0) {
            plan = response.substring(0, actionStart).trim();
        } else {
            plan = response.trim();
        }
    } else if (!plan) {
        plan = actions.length > 0 ? 'Direct action plan' : response.trim();
    }

    // Default actions if none found
    if (actions.length === 0) {
        console.log('[PARSER] No valid actions, defaulting to: right, right, a');
        actions = ['right', 'right', 'a'];
    }

    console.log('[PARSER] Plan:', plan.substring(0, 80));
    console.log('[PARSER] Actions:', actions);

    return { plan, actions };
}

File: tesserack-local-ai/agent/action_parser.py
"""Parse LLM responses into game actions."""

import re


VALID_BUTTONS = {"a", "b", "start", "select", "up", "down", "left", "right"}


def parse_response(response: str) -> tuple[str, list[str]]:
    """Parse LLM response into reasoning and actions.

    Args:
        response: Raw LLM output

    Returns:
        Tuple of (reasoning, list of button names)
    """
    print(f"[LLM RAW] {response[:500]}")  # Log first 500 chars

    reasoning = ""
    actions = []

    # Look for PLAN: line
    plan_match = re.search(r'PLAN:\s*(.+?)(?:\n|$)', response, re.IGNORECASE)
    if plan_match:
        reasoning = plan_match.group(1).strip()

    # Look for ACTIONS: line (plural, for batch)
    match = re.search(r'ACTIONS?:\s*(.+?)(?:\n|$)', response, re.IGNORECASE)
    if match:
        action_str = match.group(1)
        # Parse button sequence
        buttons = re.split(r'[,\s]+', action_str.lower())
        actions = [b.strip() for b in buttons if b.strip() in VALID_BUTTONS]

    # If no plan found, use everything before ACTIONS as reasoning
    if not reasoning:
        action_start = re.search(r'actions?:', response, re.IGNORECASE)
        if action_start:
            reasoning = response[:action_start.start()].strip()
        else:
            reasoning = response.strip()

    # Default to exploration if no valid actions found
    if not actions:
        print("[PARSER] No valid actions found, defaulting to: right, right, a")
        actions = ["right", "right", "a"]

    print(f"[PARSER] Plan: {reasoning[:100]}")
    print(f"[PARSER] Actions: {actions}")

    return reasoning, actions
